### DISCUSSION TITLE
UE5 Plugin and Weird Result About node_pcdata->set_value

### DISCUSSION THREAD

**Discussion by ffreality** [zeux/pugixml#578] [Q&A] [answered]
I am working on PugiXML based XML Creation and Parser plugin for Unreal Engine 5. It is open source and you can access it from https://github.com/FF-Plugins-Docs/FF_PugiXml

Before starting PugiXml, I created an XML plugin with TinyXml but I got some problems with Turkish characters.

If you find it useful and correct, I can add it to Show off section, too.

**My roadmap**
- XML Creation
- Parser
- Doctype detail

As mentioned on different topic, I added `pugixml.hpp and .cpp` to the source.
So, if there is no additional build necessity (like NDK), you can use it as crossplatform.
https://github.com/zeux/pugixml/issues/207

But I have a problem about child node's value.
According to these links `node_pcdata` can get a text value.
https://pugixml.org/docs/manual.html#dom.tree
https://github.com/zeux/pugixml/discussions/449

So, I wrote these codes 
https://github.com/FF-Plugins-Docs/FF_PugiXml/blob/main/Source/FF_PugiXml/Private/FF_PugiXmlBPLibrary.cpp#L246

but XML result comes like this.
```
<?xml version="1.0" encoding="utf-8" standalone="no"?>
<UE5_PUGIXML xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <COMPANIES><Company_VR>Frozen Forest</Company_VR>
    <Company_Edu>Talkido</Company_Edu>
    <Company_Web>Monolib</Company_Web>
  </COMPANIES>
</UE5_PUGIXML>
```

and I want my result like this
```
<?xml version="1.0" encoding="utf-8" standalone="no"?>
<UE5_PUGIXML xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <COMPANIES>
    <Company_VR>Frozen Forest</Company_VR>
    <Company_Edu>Talkido</Company_Edu>
    <Company_Web>Monolib</Company_Web>
  </COMPANIES>
</UE5_PUGIXML>
```
this is my blueprint nodes.
https://blueprintue.com/blueprint/-vqs-bqa/

As you can see on my result, calling `PugiXml - Add Node` for the first time adds node to right beside parent node.
but after second calling, function can add nodes to bottom lines.
This problem doesn't occurs if I don't add any value/node_pcdata

Is there any suggestion for that ?

**Comment by zeux** [comment, marked as answer]
I would recommend checking if `NodeValue` is empty before running this line:

	Out_Node->Node.append_child(node_pcdata).set_value(TCHAR_TO_UTF8(*NodeValue));

What happens is that your COMPANIES node has a first child that is pcdata (with empty value). The node formatting code disables automatic indentation in a condition like this because it assumes that the pcdata child will have some custom formatting of its own.

So as long as this line doesn't execute for COMPANIES you should see consistent indentation for all child nodes.

**Comment by ffreality** [comment]
You rock, sir !
